convoy/batch.py

```diff
@@ -4471,41 +4470,75 @@ def _construct_task(
                             ),
                             '-npernode {}'.format(processes_per_node)
                         ])
-                    mpi_opts.extend(task.multi_instance.mpi.options)
+                if is_singularity:
+                    mpi_singularity_cmd = 'singularity {} {} {} {}'.format(
+                        task.singularity_cmd,
+                        ' '.join(task.run_options),
+                        task.singularity_image,
+                        task.command)
                     mpi_command = 'mpirun {} {}'.format(
                         ' '.join(mpi_opts),
-                        'singularity {} {} {} {}'.format(
-                            task.singularity_cmd,
-                            ' '.join(task.run_options),
-                            task.singularity_image,
-                            task.command
-                        )
+                        mpi_singularity_cmd
+                    )
+                else:
+                    if task.multi_instance.mpi.runtime == 'openmpi':
+                        mpi_opts.extend(['--allow-run-as-root'])
+                    docker_mpirun_cmd = 'mpirun {} {}'.format(
+                        ' '.join(mpi_opts),
+                        task.command)
+                    docker_exec_cmds = []
+                    pre_exec_cmd = task.multi_instance.pre_execution_command
+                    if util.is_not_empty(pre_exec_cmd):
+                        docker_exec_cmds.append(pre_exec_cmd)
+                    docker_exec_cmds.append(docker_mpirun_cmd)
+                    mpi_command = 'docker exec {} {} {}'.format(
+                        ' '.join(task.docker_exec_options),
+                        task.name,
+                        util.wrap_commands_in_shell(
+                            docker_exec_cmds, windows=is_windows)
                     )
             else:
```

Comment:
add in-line comment after else regarding what conditional this refers to since the block is large